File: src/net/dinkla/raytracer/math/Normal.scala
package net.dinkla.raytracer.math

///*
//
//
///**
// * Created by Dinkla on 10.07.2015.
// */
//abstract class Normal[@specialized(Float, Double) F](x: F, y: F, z: F)(implicit numType: Fractional[F]) extends Element3D[F](x, y, z)  {
//
//  import numType.mkNumericOps
//
//  def unary_-(): Normal[F]           // unary minus
//
//  //def this(v: Vector3D[F]) = this(v.normalize)
//
//
//
//  def +(n: Normal[F]): Vector3D[F]        = plus(n)    // Normal plus Normal gives a Vector
//  def plus(n: Normal[F]): Vector3D[F]     // Normal plus Normal gives a Vector
//
//  def -(n: Normal[F]): Vector3D[F]        = plus(-n)    // Point minus Point gives a Vector
//  def minus(n: Normal[F]): Vector3D[F]    = plus(-n)    // Point minus Point gives a Vector
//
//  def toVector: Vector2D[F]
//
//}
//
//object Normal {
//
//  val RIGHT: Normal = new Normal(1, 0, 0)
//  val LEFT: Normal = new Normal(-1, 0, 0)
//  val UP: Normal = new Normal(0, 1, 0)
//  val DOWN: Normal = new Normal(0, -1, 0)
//  val FRONT: Normal = new Normal(0, 0, 1)
//  val BACK: Normal = new Normal(0, 0, -1)
//  val ZERO: Normal = new Normal(0, 0, 0)
//
//}*/

class Normal(x: Float, y: Float, z: Float) extends Vector3DF(x, y, z){

  def this(v: Vector3DF) = this(v.x, v.y, v.z)
  def this(p0: Point3DF, p1: Point3DF, p2: Point3DF) = this(new Vector3DF((p1 - p0).cross(p2 - p0).normalize))

  override def normalize: Normal = {
    val v = super.normalize
    new Normal(v.x, v.y, v.z)
  }

  // unary minus
  override def negate(): Normal = new Normal(-x, -y, -z)

}

object Normal {
  val RIGHT: Normal = new Normal(1, 0, 0)
  val LEFT: Normal = new Normal(-1, 0, 0)
  val UP: Normal = new Normal(0, 1, 0)
  val DOWN: Normal = new Normal(0, -1, 0)
  val FRONT: Normal = new Normal(0, 0, 1)
  val BACK: Normal = new Normal(0, 0, -1)
  val ZERO: Normal = new Normal(0, 0, 0)
}

//

File: src/commonMain/kotlin/net/dinkla/raytracer/math/Normal.kt
package net.dinkla.raytracer.math

import kotlin.math.sqrt

data class Normal(
    val x: Double,
    val y: Double,
    val z: Double,
) {
    operator fun plus(normal: Normal) = Vector3D(x + normal.x, y + normal.y, z + normal.z)

    operator fun times(s: Double) = Vector3D(s * x, s * y, s * z)

    operator fun unaryMinus() = Normal(-x, -y, -z)

    infix fun dot(v: Vector3D): Double = x * v.x + y * v.y + z * v.z

    fun normalize(): Normal {
        val len = length()
        return Normal(x / len, y / len, z / len)
    }

    private fun sqrLength(): Double = x * x + y * y + z * z

    fun length(): Double = sqrt(sqrLength())

    fun toVector3D() = Vector3D(x, y, z)

    companion object {
        fun create(v: Vector3D): Normal {
            val n = v.normalize()
            return Normal(n.x, n.y, n.z)
        }

        fun create(
            p0: Point3D,
            p1: Point3D,
            p2: Point3D,
        ): Normal {
            val n = ((p1 - p0) cross (p2 - p0)).normalize()
            return Normal(n.x, n.y, n.z)
        }

        val RIGHT = Normal(1.0, 0.0, 0.0)
        val LEFT = Normal(-1.0, 0.0, 0.0)
        val UP = Normal(0.0, 1.0, 0.0)
        val DOWN = Normal(0.0, -1.0, 0.0)
        val ONE = Normal(1.0, 1.0, 1.0)
        val FORWARD = Normal(0.0, 0.0, 1.0)
        val BACKWARD = Normal(0.0, 0.0, -1.0)
        val ZERO = Normal(0.0, 0.0, 0.0)
    }
}
